classifier_and_summarizer/summarizer/text_media_matcher.py

```diff
@@ -0,0 +1,173 @@
+        self.element_index = element_index
+        self.element_text = element_text
+
+
+class TextMediaMatcher:
+    ''' Class containing utilties to find the Text-Media Matching
+    Finds the stable matching by applying the Gale Shalpley algorithm
+    It accepts a list of objects of type ElementWithIndex
+    '''
+
+    def __init__(self, summary_sentences,
+                 media_descriptions, media_attributes):
+
+        self.summary_sentences = summary_sentences
+        self.media_descriptions = media_descriptions
+        self.media_attributes = media_attributes
+        self.set_size = len(media_descriptions)  # size of each set
+
+        self.sentence_embedding_model = SentenceTransformer(
+            'bert-base-nli-stsb-mean-tokens')
+
+        # similarity matrix : will be used for sorting preferences
+        # similarity score between ith summary_sentence and jth media
+        # description
+        self.similarity_matrix = [
+            [-1 for i in range(self.set_size)] for j in range(self.set_size)]
+
+        self.sentence_preference_for_media = [
+            list(range(self.set_size)) for i in range(self.set_size)]
+        self.media_preference_for_sentence = [
+            list(range(self.set_size)) for i in range(self.set_size)]
+
+    def find_stable_matching(self):
+        self._form_preference_matrix()
+        return self._gale_shapley_matching()
+
+    def _embed_text(self):
+        '''finds the vector embeddings for necessary text'''
+        self.summary_sentence_embeddings \
+            = self.sentence_embedding_model.encode(
+                [element.element_text for element in self.summary_sentences])
+
+        self.media_description_embeddings \
+            = self.sentence_embedding_model.encode(
+                [element.element_text for element in self.media_descriptions])
+
+        self.media_attribute_embeddings \
+            = self.sentence_embedding_model.encode(
+                [element.element_text for element in self.media_attributes])
+
+    def _cosine_similarity(self, embedding_A, embedding_B):
+        '''returns the cosine similarity between two vectors'''
+        return 1.0 - cosine_distance(embedding_A, embedding_B)
+
+    def _similarity_score(self, sentence_index, media_index):
+        sentence_similarity_score \
+            = (1.0 + max(
+                self._cosine_similarity(
+                    self.summary_sentence_embeddings[sentence_index],
+                    self.media_description_embeddings[media_index]
+                ),
+                self._cosine_similarity(
+                    self.summary_sentence_embeddings[sentence_index],
+                    self.media_attribute_embeddings[media_index]
+                )
+            )
+            )
+
+        distance_score \
+            = 1.0 + abs(
+                self.media_attributes[media_index].element_index
+                - self.summary_sentences[sentence_index].element_index
+            )
```

Comment:
It is more or less a safety measure to prevent division by zero. Ideally two elements shouldn't have the same index but while parsing the webpage if two elements lie in the same row i'm guessing it might have the same index . Thus the difference would be 0 and it would invoke a ZeroDivisionError.

Reply:
Ah, makes sense. Thanks for clarifying!

Reply:
Probably worth adding a comment, indicating why the 1 is added